runtime/src/transaction_cost.rs

```diff
@@ -0,0 +1,83 @@
+use solana_sdk::pubkey::Pubkey;
+
+const MAX_WRITABLE_ACCOUNTS: usize = 256;
+
+// costs are stored in number of 'compute unit's
+#[derive(Debug)]
+pub struct TransactionCost {
+    pub writable_accounts: Vec<Pubkey>,
+    pub signature_cost: u64,
+    pub write_lock_cost: u64,
+    pub data_bytes_cost: u64,
+    pub builtins_execution_cost: u64,
+    pub bpf_execution_cost: u64,
+    pub account_data_size: u64,
+    pub is_simple_vote: bool,
+}
+
+impl Default for TransactionCost {
+    fn default() -> Self {
+        Self {
+            writable_accounts: Vec::with_capacity(MAX_WRITABLE_ACCOUNTS),
+            signature_cost: 0u64,
+            write_lock_cost: 0u64,
+            data_bytes_cost: 0u64,
+            builtins_execution_cost: 0u64,
+            bpf_execution_cost: 0u64,
+            account_data_size: 0u64,
+            is_simple_vote: false,
+        }
+    }
+}
+
+#[cfg(test)]
+impl PartialEq for TransactionCost {
+    fn eq(&self, other: &Self) -> bool {
+        fn to_hash_set(v: &[Pubkey]) -> std::collections::HashSet<&Pubkey> {
+            v.iter().collect()
+        }
+
+        self.signature_cost == other.signature_cost
+            && self.write_lock_cost == other.write_lock_cost
+            && self.data_bytes_cost == other.data_bytes_cost
+            && self.builtins_execution_cost == other.builtins_execution_cost
+            && self.bpf_execution_cost == other.bpf_execution_cost
+            && self.account_data_size == other.account_data_size
+            && self.is_simple_vote == other.is_simple_vote
+            && to_hash_set(&self.writable_accounts) == to_hash_set(&other.writable_accounts)
+    }
+}
+
+#[cfg(test)]
+impl Eq for TransactionCost {}
+
+impl TransactionCost {
+    pub fn new_with_capacity(capacity: usize) -> Self {
+        Self {
+            writable_accounts: Vec::with_capacity(capacity),
+            ..Self::default()
+        }
+    }
+
+    pub fn new_with_default_capacity() -> Self {
```

Comment:
this pattern seems a bit odd to me, just use default()?

Reply:
clippy isn't happy if use `default` in cost_model.rs ln 35 above, for `clippy::field-reassign-with-default`. Yet I don't want to expose `MAX_WRITABLE_ACCOUNTS` to outside since this is a rather arb number.

Reply:
chk, leave it for follow-up.

It seems a consequence of the cost models implementation of `calculate_cost`; i.e. creates a default and then mutates the values.
Not clear to me why we chose that pattern over something more along these lines:
```rust
    pub fn calculate_cost(
        transaction: &SanitizedTransaction,
        feature_set: &FeatureSet,
    ) -> TransactionCost {
        let signature_cost = Self::get_signature_cost(transaction);
        let (write_lock_cost, writable_accounts) = Self::get_write_lock_cost(transaction);
        let (builtins_execution_cost, bpf_execution_cost, data_bytes_cost) =
            Self::get_transaction_cost(transaction, feature_set);

        TransactionCost {
            writable_accounts,
            signature_cost,
            write_lock_cost,
            data_bytes_cost,
            builtins_execution_cost,
            bpf_execution_cost,
            account_data_size: Self::calculate_account_data_size(transaction),
            is_simple_vote: transaction.is_simple_vote_transaction(),
        }
    }
```

w/ something like this we could potentially even remove the new/defaults.